src/webpage/src/stats/mod.rs

```diff
@@ -0,0 +1,110 @@
+use chrono::{DateTime, Utc};
+use ringbuffer::RingBuffer;
+use serde::Deserialize;
+
+use crate::messages::FieldInfo;
+
+pub mod drivers_stats;
+pub mod hub_messages_stats;
+pub mod hub_stats;
+pub mod stats_frequency;
+
+pub type SystemId = u8;
+pub type ComponentId = u8;
+pub type MessageId = u32;
+
+#[derive(Debug, Clone, Default, Deserialize)]
+pub struct StatsInner<B, M, D> {
+    pub last_message_time_us: u64,
+    pub bytes: B,
+    pub messages: M,
+    pub delay_stats: D,
+}
+
+impl StatsInner<ByteStatsHistorical, MessageStatsHistorical, DelayStatsHistorical> {
+    fn update(
+        &mut self,
+        now: DateTime<Utc>,
+        sample_stats: StatsInner<ByteStatsSample, MessageStatsSample, DelayStatsSample>,
+    ) {
+        // Update ByteStats
+        self.bytes
+            .total_bytes
+            .history
+            .push((now, sample_stats.bytes.total_bytes));
+        self.bytes
+            .bytes_per_second
+            .history
+            .push((now, sample_stats.bytes.bytes_per_second));
+        self.bytes
+            .average_bytes_per_second
+            .history
+            .push((now, sample_stats.bytes.average_bytes_per_second));
+
+        // Update MessageStats
+        self.messages
+            .total_messages
+            .history
+            .push((now, sample_stats.messages.total_messages));
+        self.messages
+            .messages_per_second
+            .history
+            .push((now, sample_stats.messages.messages_per_second));
+        self.messages
+            .average_messages_per_second
+            .history
+            .push((now, sample_stats.messages.average_messages_per_second));
+
+        // Update DelayStats
+        self.delay_stats
+            .delay
+            .history
+            .push((now, sample_stats.delay_stats.delay));
+        self.delay_stats
+            .jitter
+            .history
+            .push((now, sample_stats.delay_stats.jitter));
+
+        self.last_message_time_us = sample_stats.last_message_time_us;
+    }
+}
+
+#[derive(Debug, Clone, Default, Deserialize)]
+pub struct ByteStatsSample {
```

Comment:
Why not getting it from the lib ?

Reply:
It was the original plan, but the Implementations fail to compile for wasm32, and the change to organize the lib to make that easy is quite massive... I can do that if we prefer.